nifi-versioned-components/pom.xml

```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi</artifactId>
+        <version>1.14.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>nifi-versioned-components</artifactId>
+
+    <profiles>
+        <profile>
+            <id>nifi-versioned-components_jigsaw</id>
+            <activation>
+                <jdk>(1.8,)</jdk>
+            </activation>
+            <dependencies>
+                <dependency>
+                    <groupId>javax.xml.bind</groupId>
+                    <artifactId>jaxb-api</artifactId>
+                </dependency>
+            </dependencies>
+        </profile>
+    </profiles>
+
+    <!-- This module should kept to having no dependencies -->
+    <dependencies>
+        <dependency>
+            <groupId>io.swagger</groupId>
+            <artifactId>swagger-annotations</artifactId>
+            <version>1.5.16</version>
+            <scope>provided</scope>
```

Comment:
Is there a specific reason this is marked as a `provided` dependency?  This places a burden on referencing modules to include the dependency explicitly as a runtime dependency when needed.

Reply:
I discussed this dependency issue with @kevdoran and he suggested this -imo- neat idea.

The `provided` scope is to _not_ introduce the swagger dependency to _nifi-api_ build time. With this we can avoid having it on the root classpath.

The run-time burden falls to the referring modules indeed but that shouldn't be a big issue. nifi-registry already had this obligation and the nifi-framework bundle also depends on swagger-annotations.
Considering only 2 high-level api annotations are used in nifi-versioned-components I think this is the best compromise we can get.

As for why even add this change _here_. The change in this PR was originally part of https://github.com/apache/nifi/pull/5118 and I'm now separating it from the Flow Analysis change as you suggested.
So the Flow Analysis change would follow this one soon. And I'd rather have this nifi-api -> nifi-versioned-components added in this one rather than that one as it has more general - we make those classes part of the public API. That is the purpose of this new module in the first place.

Reply:
Thanks for the reply @tpalfy.

Having an additional build dependency is not a problem, but having an additional runtime dependency is the concern.  If the Swagger annotations did not have a runtime retention policy, then it seems like this approach could work.  However, if the class loader needs to have access to the Swagger annotations at runtime, then this creates issues for extension components.  Having a runtime dependency would require any extension components that access these classes to include the Swagger annotations dependency, which is a problem.  As it stands, this impacts not just the nifi-framework bundle, but extension bundles as well, which might want to leverage the versioned components.

Unless I am missing something, it seems that we need to look at an approach that avoids introducing new transitive runtime dependencies into `nifi-api`.

Perhaps @markap14 can provide additional insight on the implications.

Reply:
Maybe it makes sense to just not include nifi-versioned-components in nifi-api, to avoid the transitive dependency issue? Do they have to be combined? If not, anyone who need the nifi-versioned-components data model can depend on that and deal with the transitive dependency issue.

Reply:
I was thinking along the same lines @kevdoran.  I believe these changes will need to be part of the work necessary to implement the Flow Analysis capability, but if this PR avoids introducing `nifi-versioned-components` as a dependency in `nifi-api`, the issue with Swagger annotations could be addressed separately.

One potential solution discussed offline was moving back to [Enunciate](http://enunciate.webcohesion.com/) for generating REST API documentation from Java documentation comments, as opposed to the Swagger annotations.  With the understanding that more work would be involved to evaluate that option, it could be handled in a separate PR and this one could move forward with changing the dependencies in `nifi-api` for now.

Reply:
Yeah, I am not aware of future plans for this data model. I just figured that of course every existing use of the versioned components data model must be OK with that transitive dependency, so why not just separate them from nifi-api... Of course that might not be the case forever, but given the motivation is code reuse, I'm just trying to avoid a solution that involves duplicating this data model to have one annotated version and one non-annotated.

The swagger annotations in Registry are not just used for REST API documentation. The swagger.json file that is generated from these annotations are used in three places: 

- To generate the HTML REST API doc (could be done via Enunciate)
- By a self-hosted Swagger UI that is dynamically generated based on the contents of swagger.json
- In clients that use swagger-codegen to generate REST API client sdks from this spec. Currently this includes [nipyapi](https://github.com/Chaffelson/nipyapi), but I believe there was also talk of replacing much of the nifi-registry-web-ui data model with TypeScript generated from the REST API swagger spec.

I'm not opposed to having a separate DTO model that can support annotations and internal model of POJOs, and mappers/converters between them, if that is the only way to satisfy current and future needs.

Reply:
I'll remove the nifi-api -> nifi-versioned-components dependency then.
When I open a new PR for Flow Analysis, I'll have nifi-framework-core depend on it as well.
As suggested, later when an extension needs the VersionedComponent classes, it needs to introduce this dependency as well.
It doesn't seem to be too hard to change later if we need it and now we can move forward.

Reply:
Thanks for the reply @tpalfy, I understand the dependency change needed for Flow Analysis.  It is helpful to get this refactor completed separately, so that the review of the Flow Analysis can be more focused.  In either case, the main question comes to back to introducing the runtime dependency on `swagger-annotations` to `nifi-api`.

It sounds like avoiding the new runtime dependency in `nifi-api` would involve more work introducing a new model architecture that does not include the Swagger annotations.